File: src/main/java/macro303/keschet/pieces/Piece.java
package macro303.keschet.pieces;

import macro303.board_game.Colour;
import macro303.board_game.item.Item;
import macro303.keschet.Direction;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

/**
 * Created by Macro303 on 2018-02-08.
 */
public abstract class Piece implements Item {
	private final int maxDistance;
	@NotNull
	private final String symbol;
	@NotNull
	private final Direction[] validDirections;
	@NotNull
	private Colour teamColour;

	protected Piece(@NotNull Colour teamColour, int maxDistance, @NotNull String symbol, @NotNull Direction[] validDirections) {
		this.teamColour = teamColour;
		this.maxDistance = maxDistance;
		this.symbol = symbol;
		this.validDirections = validDirections;
	}

	protected Piece(int maxDistance, @NotNull String symbol, @NotNull Direction[] validDirections) {
		this(Colour.RESET, maxDistance, symbol, validDirections);
	}

	@NotNull
	public Colour getTeamColour() {
		return teamColour;
	}

	public void setTeamColour(@NotNull Colour teamColour) {
		this.teamColour = teamColour;
	}

	public int getMaxDistance() {
		return maxDistance;
	}

	@NotNull
	@Override
	public String getSymbol() {
		return teamColour.getColourCode() + symbol;
	}

	@NotNull
	public Direction[] getValidDirections() {
		return validDirections;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Piece)) return false;

		Piece piece = (Piece) o;

		if (maxDistance != piece.maxDistance) return false;
		if (!symbol.equals(piece.symbol)) return false;
		// Probably incorrect - comparing Object[] arrays with Arrays.equals
		if (!Arrays.equals(validDirections, piece.validDirections)) return false;
		return teamColour == piece.teamColour;
	}

	@Override
	public int hashCode() {
		int result = maxDistance;
		result = 31 * result + symbol.hashCode();
		result = 31 * result + Arrays.hashCode(validDirections);
		result = 31 * result + teamColour.hashCode();
		return result;
	}

	@Override
	public String toString() {
		return "Piece{" +
				"maxDistance=" + maxDistance +
				", symbol='" + symbol + '\'' +
				", validDirections=" + Arrays.toString(validDirections) +
				", teamColour=" + teamColour +
				'}';
	}
}

File: src/main/kotlin/macro303/keschet/pieces/Piece.kt
package macro303.keschet.pieces

import macro303.keschet.Board
import macro303.keschet.Console
import macro303.keschet.Direction
import org.apache.logging.log4j.LogManager
import java.util.*

abstract class Piece(var teamColour: Console.Colour, val maxDistance: Int, val symbol: String, val validDirections: Array<Direction>) {

	open fun validMovement(start: Pair<Int, Int>, end: Pair<Int, Int>): Boolean {
		val valid = validDirections.contains(Board.calculateDirection(start = start, end = end)) && maxDistance >= Board.calculateDistance(start = start, end = end)
		LOGGER.trace("boolean validMovement(Pair<Integer, Integer>, Pair<Integer, Integer>) = $valid")
		return valid
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Piece) return false

		if (teamColour != other.teamColour) return false
		if (maxDistance != other.maxDistance) return false
		if (symbol != other.symbol) return false
		if (!Arrays.equals(validDirections, other.validDirections)) return false

		return true
	}

	override fun hashCode(): Int {
		var result = teamColour.hashCode()
		result = 31 * result + maxDistance
		result = 31 * result + symbol.hashCode()
		result = 31 * result + Arrays.hashCode(validDirections)
		return result
	}

	override fun toString(): String {
		return "Piece(teamColour=$teamColour, maxDistance=$maxDistance, symbol='$symbol', validDirections=${Arrays.toString(validDirections)})"
	}

	companion object {
		private val LOGGER = LogManager.getLogger(Piece::class.java)
	}
}